### PR TITLE
Feature/change block reference

### PR DESCRIPTION
**Pull request by martinyde**
God fornøjelse.

### CODE REVIEW

File: web/modules/custom/lit/src/Controller/BooksController.php [hunk 1]
```diff
@@ -35,4 +36,13 @@ public function withoutImagesList(Request $request) {
             ],
         ];
     }
+
+    /**
+     * Redirect path when visiting /books or /bøger.
```
Comment (cableman) [inline web/modules/custom/lit/src/Controller/BooksController.php]: Missing return documentation

File: web/sites/default/services.local.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+parameters:
```
Comment (cableman) [inline web/sites/default/services.local.yml]: Local file on github?

File: config/sync/core.base_field_override.node.landing_page.promote.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+langcode: da
+status: true
+dependencies:
+  config:
+    - node.type.landing_page
+id: node.landing_page.promote
+field_name: promote
+entity_type: node
+bundle: landing_page
+label: 'Forfremmet til forside'
```
[the remark this answers is not in the document]
Comment (martinyde) [reply config/sync/core.base_field_override.node.landing_page.promote.yml]: When running danish as default language for a site, the config will us danish strings in the sync default folder.

File: config/sync/litteratursiden.settings.yml [hunk 4]
```diff
@@ -21,7 +21,7 @@ region_wells:
   footer: ''
 table_responsive: '-1'
 table: ''
-background_path: 'public://morten_pape_topbillede_2019_web_0.jpg'
+background_path: 'public://Skærmbillede 2020-01-02 kl. 08.39.00.png'
```
[the remark this answers is not in the document]
Comment (martinyde) [reply config/sync/litteratursiden.settings.yml]: This issue is not directly related to this release. We should ignore that specific yaml file when importing or better even change the way the background path is stored.

[the remark this answers is not in the document]
Comment (martinyde) [reply config/sync/litteratursiden.settings.yml]: This file is ignored on import : https://github.com/litteratursiden/litteratursiden/blob/feature/change-block-reference/config/sync/config_ignore.settings.yml

File: config/sync/user_registrationpassword.mail.yml [hunk 5]
```diff
@@ -1,3 +1,4 @@
 register_confirmation_with_pass:
-  subject: 'Kontooplysninger for [user:display-name] på [site:name]'
-  body: "[user:display-name],\r\n\r\nTak fordi du har registreret dig hos [site:name]. Du kan nu logge ind ved at klikke på dette link eller kopiere og indsætte det i din browser:\r\n\r\n[user:one-time-login-url]\r\n\r\nDette link kan kun bruges en gang til at logge ind og fører dig til en side, hvor du kan angive din adgangskode.\r\n\r\nNår du har indstillet din adgangskode, vil du kunne logge ind på [site:login-url] i fremtiden bruger:\r\n\r\nbrugernavn: [user:account-name]\r\nadgangskode: Din kode\r\n"
+  subject: ''
+  body: ''
+langcode: da
```
[the remark this answers is not in the document]
Comment (martinyde) [reply config/sync/user_registrationpassword.mail.yml]: Added to config ignore

File: web/modules/custom/lit/lit.install [hunk 6]
```diff
@@ -91,4 +95,105 @@ function lit_update_8003() {
...
+/**
+ * Change frontpage in config and translate strings.
+ */
+function lit_update_8106() {
+  $config_factory = Drupal::configFactory();
+  $config = $config_factory->getEditable('system.site');
+  $config->set('page.front', '/frontpage');
+  $config->save(TRUE);
+
+  _lit_add_translation('Books on the list','Bøger på listen', 'da');
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.install]: Added space

File: web/modules/custom/lit/lit.install [hunk 7]
```diff
@@ -91,4 +95,105 @@ function lit_update_8003() {
...
+}
+
+/**
+ * @param $source_string
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.install]: Added types and documentation

File: web/modules/custom/lit/lit.module [hunk 8]
```diff
@@ -230,8 +230,13 @@ function lit_node_links_alter(array &$links, NodeInterface $entity, array &$cont
       ];
     }
   }
+
+  // Remove comment links.
+  unset($links['comment__field_book_list_comments']);
 }
 
+
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.module]: Removed extra.

File: web/modules/custom/lit/lit.install [hunk 9]
```diff
@@ -91,4 +95,105 @@ function lit_update_8003() {
...
+    'title'       => 'Forfattere',
+    'path' => [
+      'alias' => '/forfattere',
+      'pathauto' => PathautoState::SKIP,
+    ]
+  ]);
+  $node->save();
+
+  MenuLinkContent::create([
+    'title' => 'Børnebøger',
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.install]: I believe a menu link is considered content and should follow the language used for content, i.e danish.

File: web/modules/custom/lit/lit.install [hunk 10]
```diff
@@ -91,4 +95,105 @@ function lit_update_8003() {
...
+    'link' => ['uri' => 'internal:/boerneboeger'],
+    'menu_name' => 'main',
+  ])->save();
+
+  MenuLinkContent::create([
+    'title' => 'Forfattere',
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.install]: I believe a menu link is considered content and should follow the language used for content, i.e danish.

File: web/modules/custom/lit/lit.links.menu.yml [hunk 11]
```diff
@@ -10,18 +10,6 @@ lit.content:
   menu_name: account
   weight: -7
 
-lit.books:
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/lit.links.menu.yml]: The old links are pointing to a route that calls a controller. The new links point to content created by editor, so we remove the old and make the menu links editable content

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 12]
```diff
@@ -0,0 +1,271 @@
+<?php
+
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: No!
"with one exception: files that contain a namespaced class/interface/trait, whose file name is the class name with a .php extension, and whose file path is closely related to the namespace (under PSR-4 or a similar standard), SHOULD NOT have a @file documentation block."

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+   *
+   * @throws \Exception
+   */
+  public function migrate() {
+    $broken_links = [];
+    $failedNodes = [];
+    $count = 0;
+    $disabled = 0;
+    $handledUrls = 0;
+    $batchCount = 2000;
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Batch count changed to be a count of blocklist

[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Changed variable name

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+    // Get all block entities.
+    $blockEntities = Drupal::entityQuery('block')->execute();
+    // Load all block config.
+    foreach ($blockEntities as $blockName) {
+      $blockList[] = Drupal::config('block.block.' . $blockName)->getRawData();
+    }
+
+    // Sort config by block weight.
+    usort($blockList, function($a, $b) {
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Added more comment

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+      return $a['weight'] - $b['weight'];
+    });
+
+    // Handle blocks.
+    foreach ($blockList as $blockConfig) {
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Added check for theme

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+            else {
+              $broken_links[$url][] = $blockConfig['settings']['label'];
+            }
+            $this->output()->writeln($trimmedUrl . ' (' . $count . '/' . $batchCount . ')');
+            $handledUrls = $handledUrls + 1;
+          }
+
+          // Delete the configuration.
+          $this->deleteConfig($blockConfig);
+          $count = $count + 1;
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Sure!

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    usort($blockList, function($a, $b) {
+      return $a['weight'] - $b['weight'];
+    });
+
+    // Handle blocks.
+    foreach ($blockList as $blockConfig) {
+      if ($blockConfig['status'] == FALSE) {
+        // Delete disabled blocks config from db.
+        $this->deleteConfig($blockConfig);
+        $disabled = $disabled + 1;
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Sure!

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+
+          // Delete the configuration.
+          $this->deleteConfig($blockConfig);
+          $count = $count + 1;
+        }
+
+        if ($blockConfig['id'] == 'views_block__recent_reviews_block') {
+          // Create views reference spot.
+          $block = BlockContent::create([
+            'info' => 'Nyeste anmeldelser',
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Yes!

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+          $block->save();
+          $url = Drupal::service('path.alias_manager')->getPathByAlias('/frontpage');
+          $urlExploded = explode('/', $url);
+          $node = Node::load($urlExploded['2']);
+          if (isset($node)) {
+            $node->field_block_reference[] = $block->id();
+            $node->save();
+          }
+
+          // $this->deleteConfig($blockConfig);
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Removed //

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+            'field_views_reference' => [
+              'target_id' => 'recent_reviews',
+              'display_id' => 'block_recent_reviews',
+              'argument' => null,
+              'title' => '1',
+              'data' => null
+            ]
+          ]);
+          $block->save();
+          $url = Drupal::service('path.alias_manager')->getPathByAlias('/frontpage');
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: The /frontpage is a node alias. I need the node id from this.

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+
+    fclose($fp);
+  }
+
+  /**
+   * Change incorrect url of landing pages.
+   *
+   * @param $url
+   *
+   * @return string
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Fixed

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+      fputcsv($fp, $fields, ';');
+    }
+
+    fclose($fp);
+  }
+
+  /**
+   * Change incorrect url of landing pages.
+   *
+   * @param $url
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Fixed

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+      case '/authors':
+        $url = '/forfattere';
+        break;
+    }
+    return $url;
+  }
+
+  /**
+   * Create a block reference on an entity.
+   *
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Added.

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 24]
```diff
@@ -0,0 +1,271 @@
...
+   * @param $url
+   *
+   * @return string
+   */
+  private function handleLandingPages(&$url) {
+    switch ($url) {
+      case '<front>':
+        $url = '/frontpage';
+        break;
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Added newlines

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+        }
+      }
+      $node = Node::load($nid);
+      if (isset($node)) {
+        $node->field_block_reference[] = $block->id();
+        $node->save();
+        $status = 'Success';
+      }
+      else {
+        $status = 'Missing block';
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Indeed

File: web/modules/custom/lit/src/Controller/AuthorsController.php [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+use Drupal\Core\Url;
+
+/**
+ * Controller for authors.
+ */
+class AuthorsController extends ControllerBase {
+
+  /**
+   * Redirect path when visiting /authors
+   */
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Controller/AuthorsController.php]: Added.

File: web/modules/custom/lit/src/Controller/BooksController.php (hunk 1 above)
Comment (martinyde) [reply web/modules/custom/lit/src/Controller/BooksController.php]: Added.

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+        }
+      }
+
+      if ($count > $batchCount) {
+         break;
+      }
+    }
+
+    // Create csv with broken links.
+    $this->createBrokenLinksCsv($broken_links);
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Changed

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+    Drupal::state()->set('block_migration_status', 'migrated');
+  }
+
+  /**
+   * Create CSV with broken links.
+   *
+   * @param $broken_links
+   */
+  private function createBrokenLinksCsv($broken_links) {
+    $fp = fopen('file.csv', 'w');
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: This is ok.

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+    $this->output()->writeln('--- Handled urls ---');
+    $this->output()->writeln($handledUrls);
+
+    Drupal::state()->set('block_migration_status', 'migrated');
+  }
+
+  /**
+   * Create CSV with broken links.
+   *
+   * @param $broken_links
```
[the remark this answers is not in the document]
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Added.

File: web/modules/custom/lit/lit.install [hunk 30]
```diff
@@ -167,15 +167,17 @@ function lit_update_8106() {
...
 }
 
 /**
- * @param $source_string
+ * Translate a string
+ *
+ * @param $source_string string
  *   The string to translate.
- * @param $translated_string
+ * @param $translated_string string
```
Comment (cableman) [inline web/modules/custom/lit/lit.install]: Should be `@param TYPE $var`

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php [hunk 31]
```diff
@@ -180,10 +186,13 @@ private function handleLandingPages(&$url) {
   /**
    * Create a block reference on an entity.
    *
-   * @param $nid
+   * @param $nid integer
```
Comment (cableman) [inline web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Wrong order and it's `int` in PHP.

File: web/modules/custom/lit/lit.install (hunk 30 above)
Comment (martinyde) [reply web/modules/custom/lit/lit.install]: Fixed

File: web/modules/custom/lit/src/Commands/ConfigToBlockReference.php (hunk 31 above)
Comment (martinyde) [reply web/modules/custom/lit/src/Commands/ConfigToBlockReference.php]: Fixed

**Pull request merged**